### PR TITLE
Update CUDA out of memory mesage with private pool info

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124673

Fixes https://github.com/pytorch/pytorch/issues/121932

cc @mcarilli @ezyang @eellison @peterbell10

### CODE REVIEW

**Comment by isuruf** [comment]
cc @eellison @peterbell10

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -1125,6 +1125,19 @@ class DeviceCachingAllocator {
...
+      size_t allocated_in_private_pools = 0;
+      auto get_size_block = [](const BlockPool& pool) {
+        size_t res = 0;
+        for (const auto& block : pool.blocks) {
+          res += block->size;
+        }
+        return res;
+      };
+      for (const auto& p : graph_pools) {
+        allocated_in_private_pools += get_size_block(p.second->large_blocks);
```
Comment (peterbell10) [inline c10/cuda/CUDACachingAllocator.cpp:1137]: This likely needs to happen before the mutex unlock on line 1119.

File: c10/cuda/CUDACachingAllocator.cpp [hunk 2]
```diff
@@ -1157,9 +1170,12 @@ class DeviceCachingAllocator {
...
           proc_info,
           "Of the allocated memory ",
-          format_size(allocated_bytes),
-          " is allocated by PyTorch, and ",
-          format_size(reserved_bytes - allocated_bytes),
+          format_size(allocated_bytes + allocated_in_private_pools),
+          " is allocated by PyTorch, with ",
+          format_size(allocated_in_private_pools),
+          " allocated in private pools, and ",
+          format_size(
```
Comment (eellison) [inline c10/cuda/CUDACachingAllocator.cpp:1177]: I don't think we should print this if there aren't any allocated private pools. Also, maybe worth explicitly mentioning cudagraphs here? for e.g. users doing `torch.compile(mode='reduce-overhead')` they might not make the connection

File: c10/cuda/CUDACachingAllocator.cpp [hunk 3]
```diff
@@ -1111,6 +1111,26 @@ class DeviceCachingAllocator {
...
+      for (const auto& p : graph_pools) {
+        allocated_in_private_pools += get_size_block(p.second->large_blocks);
+        allocated_in_private_pools += get_size_block(p.second->small_blocks);
+      }
+
+      std::string private_pool_msg;
+
+      if (allocated_in_private_pools > 0) {
+        private_pool_msg = "with " + format_size(allocated_in_private_pools) +
+            " allocated in private pools (eg: CUDAGraphs), ";
```
Comment (eqy) [inline c10/cuda/CUDACachingAllocator.cpp:1131]: nit: `(e.g., CUDA Graphs)`

**Review by eellison — APPROVED** [review]
🚢

**Review by eqy — APPROVED** [review]

**Pull request closed without merging**